### ISSUE TITLE
BUG: handle inconsistencies in factorial functions and their extensions

### ISSUE THREAD

**Issue by h-vetinari** [scipy/scipy#15600]
In the context of #15351 and reviewing #15299, I noticed a bunch of inconsistencies that should be fixed IMO.

## TL;DR 

[Courtesy](https://github.com/scipy/scipy/issues/15600#issuecomment-1041258485) of @tirthasheshpatel:

* `factorial`, `factorial2`, and `factorialk` behave inconsistently for several different combinations of inputs. (for example, `factorial2` returning 0-dim arrays for scalars and, `factorial2` and `factorialk` aren't vectorized). See the first table for more details. This issue can be tacked by adding a new parameter like `legacy` which, by default, preserves the old behaviour but when false, handles the current inconsistent behavior correctly.
* Support for float and complex inputs can easily be added using [this extension](https://en.wikipedia.org/wiki/Double_factorial#Alternative_extension_of_the_multifactorial). This enhancement adds a new boolean parameter `extend='zero'|'complex'` (and perhaps `'recurrence'` down the line) that specifies whether to use the extension formula or the default formula. See the second table for more details.

## Design Space

There are several dimensions that come into play:
- scalar vs. array inputs
- for integers: approximation or exact computation (`exact=`)
- extensions to negative integers
- extensions to floats
- extensions to complex numbers
- extensions to poles (possible e.g. for `factorial2`)
- for the array-case, failing soft (e.g. inserting NaNs) or hard (raising error)

## Current Behaviour (as of 1.8.0)

| Inputs | `factorial(`<br/>`exact=False)` | `factorial(`<br/>`exact=True)` | `factorial2(`<br/>`exact=False)` | `factorial2(`<br/>`exact=True)` | `factorialk(`<br/>`exact=False)` | `factorialk(`<br/>`exact=True)` |
|-|-|-|-|-|-|-|
| Scalar positive integer | float | int | !! _**0-dim<br/>float array**_ !! | int | `NotImpl` `Error` | int |
| Scalar negative integer | 0.0 | 0 | !!&nbsp;`array(0.)`&nbsp;!! |  0 | `NotImpl` `Error` | 0 |
| Scalar positive float | float | `ValueError` (but text misleading) | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |
| Scalar negative float | 0.0 | 0 | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |
| Scalar complex | `TypeError` | `TypeError` | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |
| Scalar NaN | NaN | NaN | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |
| Positive integer in array | float | int | float | !!&nbsp;`TypeError`&nbsp;!! | `NotImpl` `Error` | `TypeError` |
| Negative integer in array | 0.0 | 0 | 0.0 | !! `TypeError` !! | `NotImpl` `Error` | `TypeError` |
| Positive float in array | float | `ValueError` (but text misleading) | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |
| Negative float in array | 0.0 | 0 | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |
| Complex in array | `TypeError` | `TypeError` | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |
| NaN in array | NaN | NaN | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |
| Empty array | `array([])` | `IndexError` | `TypeError` | `TypeError` | `NotImpl` `Error` | `TypeError` |

Note that:
- `factorial2(<scalar>, exact=False)` returns a zero-dimensional array instead of a scalar.
- All the `TypeErrors` are IMO implementation oversights (and in case of "!!", against documentation) - i.e. we should provide better error messages.
- There is a full [generalization](https://en.wikipedia.org/wiki/Double_factorial#Alternative_extension_of_the_multifactorial) of `factorialk` which is basically just a fraction of two gamma functions, which would work for all complex n & k (up to poles), as well as arrays.

## Extension Issues

In addition to this hodgepodge of behaviour, some people would like to have the float/complex extensions of these functions (e..g. https://github.com/scipy/scipy/pull/15349). I tend to agree, but that raises another problem:

#### Complex extensions are incompatible with current inputs ≤0

First off, everything that's 0 currently on the negative axis gets a different value. But also note the "≤"; namely, 0!! is 1 in the regular definition, but ~0.79=sqrt(2/pi) in the [extension](https://en.wikipedia.org/wiki/Double_factorial#Complex_arguments).

#### Extensions based on recurrence are incompatible with current status as well as complex extension

This came up for example in #15299. It's [possible](https://en.wikipedia.org/wiki/Double_factorial#Negative_arguments) (for example) to invert the recurrence relation n!! = n * (n-2)!!, and so extend the double-factorial to negative integers. The resulting value would be neither 0 (currently) nor the pole in the complex extension.

## Design Considerations

If we (or at least I) were designing from scratch:
- Changing from the `exact=False` to `exact=True` should _never_ fundamentally change the result (e.g. going from 0 to non-zero, or casting inputs to integers as seen in #15351).
  - As such, the exact keyword only really matters to positive scalar integers; in all other cases, it could be retired, resp. raise, or simply be turned into a no-op (cf. #15359)
  - Open question whether it's worth to properly implement `exact=True` for array-case (currently, only `factorial` does).
- Complex extensions  (together with non-positive floats) should be behind a keyword, roughly because:
  - 0-for-negative-values is a valid approach, and has been in scipy ~forever.
  - Since the values being produced are incompatible (for certain inputs), the complex extension should be explicitly opt-in.
- Using recurrence relations for extending into poles is an open question; not sure if worth it, but if so, probably with another ~keyword~ value for the `extend` keyword.
- Use the most general approximation, except when alternatives with better accuracy / performance exist.
- Probably need to think about dtype casting rules for empty arrays (keep dtype? always float? ...)

## Possible Future Behaviour (note `extend=`, not `exact=`)

| Inputs | `factorial(`<br/>`extend=`<br/>`'zero')` | `factorial(`<br/>`extend=`<br/>`'complex')` | `factorial2(`<br/>`extend=`<br/>`'zero')` | `factorial2(`<br/>`extend=`<br/>`'complex')` | `factorialk(`<br/>`extend=`<br/>`'zero')` | `factorialk(`<br/>`extend=`<br/>`'complex')` |
|-|-|-|-|-|-|-|
| Scalar positive integer&nbsp;w/ `exact=True` | n! | n! | n!! | n!! | n!(k) | n!(k) |
| Scalar positive integer | Γ(n+1) | Γ(n+1) | (1) | (1) | (2) | (2) |
| Scalar negative integer / poles | 0.0 | NaN | 0.0 | NaN | 0.0 | NaN |
| Scalar positive float | Γ(n+1) | Γ(n+1) | (1) | (1) | (2) | (2) |
| Scalar negative float | 0.0 | Γ(n+1) | 0.0 | (1) | 0.0 | (2) |
| Scalar complex | `ValueError` | Γ(n+1) | `ValueError` | (1) | `ValueError` | (2) |
| Scalar NaN | NaN | NaN | NaN | NaN | NaN | NaN |
| Positive integer in array | Γ(n+1) | Γ(n+1) | (1) | (1) | (2) | (2) |
| Negative integer in array | 0.0 | NaN | 0.0 | NaN | 0.0 | NaN |
| Positive float in array | Γ(n+1) | Γ(n+1) | (1) | (1) | (2) | (2) |
| Negative float in array | 0.0 | Γ(n+1) | 0.0 | (1) | 0.0 | (2) |
| Complex in array | `ValueError` | Γ(n+1) | `ValueError` | (1) | `ValueError` | (2) |
| NaN in array | NaN | NaN | NaN | NaN | NaN | NaN |
| Empty array | `array([])` | `array([])` | `array([])` | `array([])` | `array([])` | `array([])` |

The formulas in question are:
* (1): z!! = 2^((z-1)/2) Γ(z/2 + 1) / Γ(1/2 + 1)
* (2): z!(α) =  α^((z-1)/α) Γ(z/α + 1) / Γ(1/α + 1)

I'd very much welcome feedback on this - there are some trade-offs to be made (e.g. whether negative floats should also raises without `extend='complex'` or whether the errors in the array-case should just map to NaN; also there are different approximations for n!! in use [currently](https://github.com/scipy/scipy/blob/v1.8.0/scipy/special/_basic.py#L2411-L2412)), but I think the above would be reasonable. In particular, the scalar and array case behave the same way.

## Implementation Plan

I would propose to introduce another keyword (sorry...) `legacy`, defaulting to `True`, but implement `legacy=False` with all the corrected behaviour. Where there's any planned change in behaviour, we raises a deprecation warning recommending to switch to `legacy=False`. After 2 releases, we switch the default from `legacy=True` to `legacy=False`, after two more releases we raise on `legacy=True`, and after two more we remove the keyword.

Looking more closely at relevant changes in (currently non-raising) behaviour, it really only affects the zero-dimensional arrays for `factorial2(scalar_val, exact=False)`. In particular, if we let `exact=True` be a no-op for non-integers (the less "noisy" options from the point of UX), we would keep essentially all current non-raising behaviour intact.

CC @steppi @tirthasheshpatel @j-bowhay @tupui

### Edit History:

* v4 (2022-03-10): Make `exact=True` a no-op for non-integers (less noisy), following #15359. Extend deprecation period of `legacy`-keyword along the same lines as proposed in https://github.com/scipy/scipy/issues/15722.
* v3 (2022-02-16; up to rev 20): Added (current & proposed) behaviour for NaN and empty arrays; good catch @tirthasheshpatel!
* v2 (2022-02-16; up to rev 15): Added TL;DR, clarification on impact of `legacy`, changed `extend` from boolean argument to take an enum.
* v1 (2022-02-16; up to rev 10): Original

**Comment by j-bowhay** [comment]
Thank you this makes it a lot easier to follow because, as you say, there are a lot of inconsistencies. I will try to have a look over this at the weekend.

**Comment by tirthasheshpatel** [comment]
Thanks for the _incredibly detailed_ report @h-vetinari :exploding_head:! I suggest adding a TLDR; something like this if I understand the issue correctly:

1. `factorial`, `factorial2`, and `factorialk` behave inconsistently for several different combinations of inputs. (for example, `factorial2` returning 0-dim arrays for scalars). See the first table for more details. This issue can be tacked by adding a new parameter like `legacy` which, by default, preserves the old behaviour but when false, handles the current inconsistent behavior correctly.
2. Support for float and complex inputs can easily be added using [this extension](https://en.wikipedia.org/wiki/Double_factorial#Complex_arguments). This enhancement adds a new boolean parameter `extend` that specifies whether to use the extension formula or the default formula. See the second table for more details.

Can this issue be split into two: one for addressing inconsistencies and one for adding the extension?

> * `factorial2(<scalar>, exact=False)` returns a zero-dimensional array instead of a scalar.

This is somewhat common in `stats` too. Although it's an inconsistency, such outputs mostly behave like scalars (and also get coerced into one if operated with scalars). I think there are more important things we should tackle first: for example, vectorizing `factorial2` and `factorialk` (these functions throwing for array inputs is very unintuitive).

> * As such, the exact keyword only really matters to positive scalar integers; in all other cases, it could be retired, resp. raise.

I agree that raising would be nice but that would again be a backward-incompatible change. Rather, we can add a keyword, similar to `nan_policy`, that asks the user whether to coerce the inputs to integers, raise, or fill NaNs for invalid cases in array inputs (e.g. `exact_policy='raise' | 'coerce' | 'fill-nan'`). For `factorial`, the default could be `'coerce'`. For `factorial2` and `factorialk`, `raise`.

> * Complex extensions  (together with non-positive floats) should be behind a keyword

This sounds good.

> I would propose to introduce another keyword (sorry...) `legacy`, defaulting to `True`, but implement `legacy=False` with all the corrected behaviour. Where there's any planned change in behaviour, we raises a deprecation warning recommending to switch to `legacy=False`. After 2 releases, we switch the default from `legacy=True` to `legacy=False`, and after two more releases we remove the keyword.

I am not against this but maybe we can also consider an alternative approach like adding `exact_policy`?

**Comment by h-vetinari** [comment]
> Thanks for the _incredibly detailed_ report @h-vetinari 🤯!

🙃 

>  I suggest adding a TLDR; something like this if I understand the issue correctly:

Done, thanks!

> Can this issue be split into two: one for addressing inconsistencies and one for adding the extension?

Fine by me, and effectively adopted with your TL;DR.

> > * `factorial2(<scalar>, exact=False)` returns a zero-dimensional array instead of a scalar.
> 
> This is somewhat common in `stats` too. Although it's an inconsistency, such outputs mostly behave like scalars (and also get coerced into one if operated with scalars). I think there are more important things we should tackle first: for example, vectorizing `factorial2` and `factorialk` (these functions throwing for array inputs is very unintuitive).

Prioritization aside, I think we should clean this up. 0-dim arrays used to be a bit of a wild west thing, but it's getting cleaned up in lots of APIs (not just numpy/scipy, also other downstreams like cvxpy).

> > * As such, the exact keyword only really matters to positive scalar integers; in all other cases, it could be retired, resp. raise.
> 
> I agree that raising would be nice but that would again be a backward-incompatible change. Rather, we can add a keyword, similar to `nan_policy`, that asks the user whether to coerce the inputs to integers, raise, or fill NaNs for invalid cases in array inputs (e.g. `exact_policy='raise' | 'coerce' | 'fill-nan'`). For `factorial`, the default could be `'coerce'`. For `factorial2` and `factorialk`, `raise`.

I'm not a fan, to be honest. It has the same high price (another new keyword) for very little gain IMO. In particular, I think coerce is basically never the right thing, and if so, then people should get the warning/error and explicitly cast.

> > * Complex extensions  (together with non-positive floats) should be behind a keyword
> 
> This sounds good.

We'll have to do some bikeshedding on this. I proposed `extend=True|False` more or less as a stand-in, because `extend_complex` seemed too long), but if we want to leave the recurrence thing on the table, we could actually do `extend='zero'|'complex'|'recurrence'`, which to me looks even better than boolean switches.

> > I would propose to introduce another keyword (sorry...) `legacy`, defaulting to `True`, but implement `legacy=False` with all the corrected behaviour. Where there's any planned change in behaviour, we raises a deprecation warning recommending to switch to `legacy=False`. After 2 releases, we switch the default from `legacy=True` to `legacy=False`, and after two more releases we remove the keyword.
> 
> I am not against this but maybe we can also consider an alternative approach like adding `exact_policy`?

I understand that it's not attractive, but I had a closer look and added this to the OP:
> Looking more closely at relevant changes in (currently non-raising) behaviour, it really only affects the zero-dimensional arrays for `factorial2(scalar_val, exact=False)`. In particular, we could choose to let `exact=True` only raise if `extend=True` as well, keeping essentially all current non-raising behaviour intact.

**Comment by tirthasheshpatel** [comment]
Some more inconsistencies: `factorial` only works with empty arrays when `exact=False`. For `exact=True`, it fails:

```py
from scipy.special import factorial, factorial2, factorialk

factorial([])  # array([], dtype=float64)
factorial([], exact=True)  # IndexError: index -1 is out of bounds for axis 0 with size 0
```

`factorial2` fails for NaN inputs:

```py
factorial2(nan, exact=False)
# TypeError: ufunc 'bitwise_and' not supported for the input types, and the inputs could not be safely coerced to any supported
# types according to the casting rule ''safe''
```

**Comment by h-vetinari** [comment]
> Some more inconsistencies

Good points! Added to the OP. Do we also want to consider `None`?

**Comment by j-bowhay** [comment]
Another possibility mentioned in the review for #15299 is making `factorial` and `factorial2` thin wrappers around `factorialk`. I'm not sure if it is a good idea but probably worth discussing.

**Comment by h-vetinari** [comment]
> @j-bowhay: [...] making `factorial` and `factorial2` thin wrappers around `factorialk`.

That's partly what I was referring to with:
> Use the most general approximation, except when alternatives with better accuracy / performance exist.

But to do that, we need some benchmarks against the current most important code paths (haven't verified current status of factorial in `benchmarks/`), and ideally some as-tight-as-possible accuracy tests.

**Comment by j-bowhay** [comment]
> But to do that, we need some benchmarks against the current most important code paths (haven't verified current status of factorial in `benchmarks/`), and ideally some as-tight-as-possible accuracy tests.

Agreed, this was my hesitation in making them thin wrappers is a potential loss in performance and accuracy as some cases have more optimal ways of performing the calculation that don't generalise.

In `benchmarks/benchmarks/special.py` there are currently no benchmarks for any of the factorial stuff. Maybe a good first step would be to add some benchmarks?

**Comment by h-vetinari** [comment]
> Maybe a good first step would be to add some benchmarks?

Absolutely! :)

IMO we should aim basically for all the cases in the first table that don't raise (except NaN/empty array).

**Comment by h-vetinari** [comment]
Since it would be good if we can do this at the same time with the changes in `comb`, I'm tentatively milestoning this for 1.9.

**Comment by h-vetinari** [comment]
While looking at the code for this, I realized that the array-path for `factorial(exact=True)` depends on the fact that `np.unique` returns sorted unique values, and so a recent behaviour [change](https://github.com/numpy/numpy/pull/18070) in numpy has at least disabled the type-setting [logic](https://github.com/scipy/scipy/blob/v1.8.0/scipy/special/_basic.py#L2365-L2375). It doesn't seem to break too loudly though, so I'm not raising a separate issue, but thought I'd note it here.

**Comment by h-vetinari** [comment]
Also, I found that - contrary to their documentation - `factorial2` and `factorialk` return wrong values for small negative integers. I'm not planning to go through a multi-release deprecation cycle for this, and will just fix it outright to return zero, as intended/documented.

```
>>> factorial2(0)
array(1.)
>>> factorial2(-1)
array(1.)
>>> factorial2(-2)
array(0.)
```
Note `factorial2(-1)` should be 0.

```
>>> factorialk(0, k=3)
1
>>> factorialk(-1, k=3)
1
>>> factorialk(-2, k=3)
1
>>> factorialk(-3, k=3)
0
```
Same here, and it's actually dependent on k (i.e. it's falsely non-zero until `-(k-1)`).

**Comment by h-vetinari** [comment]
While working on this, I just stumbled over [`test_factorial_0d_return_type`](https://github.com/scipy/scipy/blame/v1.8.0/scipy/special/tests/test_basic.py#L1986-L1987), and thus https://github.com/scipy/scipy/issues/7400 & https://github.com/scipy/scipy/pull/10954. This shows that the 0d array-output for `factorial2` was already undesired a long time ago, and was merely overlooked in #10954.

The reason this is relevant is that if we treat it as a bugfix (similar to #10954), then we'd have no deprecations to do for the other points in this issue.

However, I wanted to tag the participants of those old discussion, because I think #10954 went too far in fixing #7400 - I don't think the factorial-functions should be changing array-inputs (even zero-dimensional ones) to scalars. IMO it should be scalar->scalar, array->array (and the latter with identical dimensions).

CC @WarrenWeckesser @person142 @rgommers @larsoner

**Comment by h-vetinari** [comment]
> Also, I found that - contrary to their documentation - `factorial2` and `factorialk` return wrong values for small negative integers. I'm not planning to go through a multi-release deprecation cycle for this, and will just fix it outright to return zero, as intended/documented.

I just realised that this statement was true for 1.8.0 (which were the sources I was looking at), but this was changed for 1.9 in https://github.com/scipy/scipy/issues/15589, https://github.com/scipy/scipy/pull/15590, https://github.com/scipy/scipy/issues/15592.

I strongly believe that the 1.8. docstrings were correct behaviour, and we should match the code, rather than document broken behaviour. CC @AtsushiSakai @tupui @j-bowhay

**Comment by h-vetinari** [comment]
Next problem - the "universal" approximations for n!! is very bad for even numbers (the odd ones work well). For example, the formula (1) from the OP, calculated to high precision with mpmath gives: 10!! = 3063.8767134830029.

However, the true value is very different: 
```
>>> factorial2(10)
3840.
```
and the current approximation actually does an extremely good job (down to a single eps even for n=100)
```
>>> factorial2(10) / float(factorial2(10, exact=True)) -1
0.0
>>> factorial2(100) / float(factorial2(100, exact=True)) - 1
2.220446049250313e-16
>>> float(factorial2(100, exact=True))
3.4243224702511973e+79
```

However, the formula picked up by @j-bowhay in #15349 seems to do a much better job also for even integers.

**Comment by h-vetinari** [comment]
> I strongly believe that the 1.8. docstrings were correct behaviour, and we should match the code, rather than document broken behaviour.

Regarding the small-negative-integers for `factorialk`, I'll note that a common [definition](https://en.wikipedia.org/wiki/Double_factorial#Definitions) does include n!!=1 for small negative `n`, but I claim that's purely for the mathematical convenience of simplifying the recursion (which is - mixing metaphors - a kind of mathematical implementation detail, but not required for an unambiguous and consistent definition**).

This makes outright changing the behaviour slightly harder, but I still think that `extend='zero'` would be by far the most sane default for negative numbers. However, if people believe it necessary, we can go through the full `legacy=` deprecation dance for that.

<details>
<summary>** alternative definition</summary>

The definition could be rewritten equivalently for all non-negative values:
```
n!(α) = n * (n - α)!(α) if n ≥ α
      = 1 if 0 ≤ n < α
      = 0 otherwise
```

</details>

**Comment by h-vetinari** [comment]
I have a WIP PR that addresses some of the issues mentioned in the OP and recent comments. It's not finished yet, and I expect that I'll have to cut it into smaller pieces, but if someone's interested, please take a look: https://github.com/scipy/scipy/pull/15841

**Comment by ilayn** [comment]
Almost in every software I used, if we have a multiple function doing similar but different versions of a functionality 5 years down the line we struggle deprecating/removing them.

I would have liked this function to be in NumPy namespace since it is a very common math operation or from standard `math` library but it is now living in `special` functions territory because reasons. 

So at least can we combine them into a single signature since they are special cases of k=1, 2 and get rid of the others eventually in the future?

**Comment by h-vetinari** [comment]
> So at least can we combine them into a single signature since they are special cases of k=1, 2 and get rid of the others eventually in the future?

I sympathise with the sentiment, though I'm not sure I agree... Aside from IMO quite well-established names (not just in scipy), the functions have some pretty subtantial differences, despite how closely related they are - e.g. `factorialk` doesn't have a well-known meromorphic continuation (meaning we cannot interpolate to floats; contrary to what I claimed in the OP), `factorial` does not permit something like `extend='recurrence'` but `factorial2` does, etc. etc.

I mean, yes, of course it's possible, but effectively it'd be three different implementations in a trenchcoat pretending to be one function.

I cannot comment on whether this belongs to numpy or scipy, but I'd be open to shepherd the PR wherever necessary.

**Comment by h-vetinari** [comment]
> I mean, yes, of course it's possible, but effectively it'd be three different implementations in a trenchcoat pretending to be one function.

Though if I can pull of the array-path of `factorial2` / `factorialk` for `exact=True` as I imagine it (based on a generalized `_exact_factorial_array`), then I think it might be possible after all to share most of the surrounding code.

**Comment by j-bowhay** [comment]
> Though if I can pull off the array-path of `factorial2` / `factorialk` for `exact=True` as I imagine it (based on a generalized `_exact_factorial_array`), then I think it might be possible after all to share most of the surrounding code.

I agree it think the not exact path is too different for each to be one function.

**Comment by ilayn** [comment]
What I mean is that from the user's perspective nobody is interested in how the code traverses the internal paths. It just needs a proper factorial function which is already buried under a `special` module whereas it is always included in the primary namespace in other (commercial/OSS) tools and unfortunately not in NumPy. 

`pinv2`, `lsim2` and so on shared also very much the same thinking back then but probably we should have been a bit more careful (20/20 hindsight). I am wondering if we are doing the same with these factorial functions

**Comment by h-vetinari** [comment]
> `pinv2`, `lsim2` and so on shared also very much the same thinking back then but probably we should have been a bit more careful (20/20 hindsight)

I'm not sure the case is fully comparable. pinv/pinv2 are doing the same thing conceptually but were using different implementations - as you wrote in #13831:

> Historically, pinv2 and pinv were using different solution methods however for a very long time already both use the same SVD based solvers. Hence duplicate code for no apparent reason in the namespace.

Here however, factorial2 does not mean "we're adding a 2 to not break the old API/behaviour", but it is a very different (though obviously related) function that has its use-cases independently from factorial.

That being said, I don't mind in principle to fold everything into `factorial` and deprecate `factorial2` / `factorialk`, though I don't think it's a huge benefit API- and especially maintenance-wise (the code will need to branch a lot to deal with different behaviours, extensibility, etc. of these functions)

**Comment by h-vetinari** [comment]
> a proper factorial function which is already buried under a `special` module whereas it is always included in the primary namespace in other (commercial/OSS) tools and unfortunately not in NumPy.

Just stumbled over this in the git-blame - seems the factorial functions used to be in an even more obscure place until https://github.com/scipy/scipy/pull/3151.

**Comment by h-vetinari** [comment]
I have an almost fully passing PR (some issues with GCC6) for all the fixes & behaviour changes in #15841. I'd appreciate if people who are interested could take a look. 🙃 

It does not yet include the complex extension, but is very thoroughly tested and should get rid of basically all the spurious TypeErrors etc. from the OP.

It also does not yet attempt to unify everything under `factorial` (cf. the most recent comments here with @ilayn).

**Comment by ilayn** [comment]
I think I should concede my comments this is indeed a quite complicated issue and I don't want to muddy it with philosophical issue. We can come back to it later and since there is no user noise around this apparently people are not too unhappy about it.

**Comment by j-bowhay** [comment]
@h-vetinari what state is your pr in? It would be nice to try get this in 1.10.0 so that the comb and factorial tidy ups end at roughly the same time.

**Comment by h-vetinari** [comment]
There was an open question (mainly for @ilayn) if we're fine with using the canonical extension for the double-factorial, which is off by a factor for even integers. I have some change requests pending but wanted to have some clarity on that before I start changing things around.

Agreed with 1.10, though that's >6 months away, so hopefully enough time.

**Comment by ilayn** [comment]
My bad. Sorry for ignoring, reading it a bit more, it seems to me that I shouldn't be the one judging this as I am not 100% on every detail for the use case. Let me ping @WarrenWeckesser and @person142 just in case they want to have a look.

**Comment by h-vetinari** [comment]
I consider this issue solved by #15841. For the open points regarding the extensions, I've opened a follow-up issue: https://github.com/scipy/scipy/issues/18409
